source/extensions/quic_listeners/quiche/platform/quic_port_utils_impl.cc

```diff
@@ -0,0 +1,47 @@
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "extensions/quic_listeners/quiche/platform/quic_port_utils_impl.h"
+
+#include "envoy/network/address.h"
+
+#include "common/common/assert.h"
+#include "common/network/utility.h"
+
+#include "test/test_common/environment.h"
+#include "test/test_common/network_utility.h"
+
+namespace quic {
+
+int QuicPickUnusedPortOrDieImpl() {
```

Comment:
This is test-only, right? I would ideally like to see a better naming or structuring to indicate code is for test vs. prod., since there are different requirements here as far as error handling and robustness. Are we restricted by the QUICHE layout?

Also, why does QUICHE do this? In Envoy, we just bind zero on the server side and then communicate the effective port OOB to the client in integration tests. Could QUICHE do this?

Reply:
> This is test-only, right? I would ideally like to see a better naming or structuring to indicate code is for test vs. prod., since there are different requirements here as far as error handling and robustness. Are we restricted by the QUICHE layout?
Talked with Harvey about a reasonable approach, such test only code can be moved to /test, and be included by a dummy impl.h file in /source. 

> Also, why does QUICHE do this? In Envoy, we just bind zero on the server side and then communicate the effective port OOB to the client in integration tests. Could QUICHE do this?
Quic e2e tests doesn't have this server->client channel to communicate the port it picks after starting listening.